pkg/gds/admin/v2/csurf.go

```diff
@@ -0,0 +1,100 @@
+package admin
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/rs/zerolog/log"
+)
+
+// Parameters and names for double-cookie submit CSRF protection
+const (
+	CSRFCookie          = "csrf_token"
+	CSRFReferenceCookie = "csrf_reference_token"
+	CSRFHeader          = "X-CSRF-TOKEN"
+)
+
+// DoubleCookie is Cross Site Request Forgery (CSRF/XSRF) protection middleware that
+// checks the presence of an X-CSRF-TOKEN header containing a cryptographically random
+// token that matches a token contained in the CSRF-TOKEN cookie in the request.
+// Because of the same-origin policy, an attacker cannot access the cookies or scripts
+// of the safe site therefore it will not be able to guess what token to put in the
+// request, thereby protecting from CSRF attacks. Note that Double Cookie CSRF
+// protection requires TLS to prevent MITM attacks.
+func DoubleCookie() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		cookie, err := c.Cookie(CSRFReferenceCookie)
+		if err != nil {
+			log.Warn().Err(err).Msg("no csrf token cookie in request")
+			c.JSON(http.StatusForbidden, ErrorResponse(ErrCSRFVerification))
+			c.Abort()
+			return
+		}
+
+		header := c.GetHeader(CSRFHeader)
+		if header, err = url.QueryUnescape(header); err != nil {
+			log.Warn().Err(err).Msg("could not unescape csrf token")
+			c.JSON(http.StatusBadRequest, ErrorResponse(err))
+			c.Abort()
+			return
+		}
+
+		if cookie != header {
```

Comment:
Great question, I just did some Googling and StackOverflow says:

> You can't set a cookie with an empty string as it will delete the cookie. From the docs: If the value argument is an empty string, or FALSE, and all other arguments match a previous call to setcookie, then the cookie with the specified name will be deleted from the remote client.

If the cookie doesn't exist then the err on L32 will go off. Header definitely can be empty string; I think the `cookie_exists` bool was my just trying to add symmetry to `header_exists` even though it will always be true if we get to this point.